File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.java
class Solution {
    private int gcd(int x, int y)
    {
        return y == 0 ? x : gcd(y, x % y);
    }

    public long gcdSum(int[] nums) {
        int n = nums.length;
        int[] pre = new int[n];
        int mx = 0;

        for(int i = 0; i < n; i++)
        {
            mx = Math.max(mx, nums[i]);
            pre[i] = gcd(nums[i], mx);
        }

        Arrays.sort(pre);
        long ans = 0;

        for(int l = 0, r = n-1; l < r; l++, r--)
            ans += gcd(pre[l], pre[r]);
        
        return ans;
    }
}

File: 3867-sum-of-gcd-of-formed-pairs/3867-sum-of-gcd-of-formed-pairs.py
class Solution:
    def gcdSum(self, nums: list[int]) -> int:
        n = len(nums)
        pre = [0] * n
        res = 0

        for i, x in enumerate(nums):
            res = max(res, x)
            pre[i] = gcd(x, res)
        
        pre.sort()
        return sum(gcd(pre[i], pre[-1-i]) for i in range(n // 2))